src/engine/profiler.cc

```diff
@@ -0,0 +1,189 @@
+/*!
+ * Copyright (c) 2015 by Contributors
+ * \file profiler.cc
+ * \brief implements profiler
+ */
+#include <dmlc/base.h>
+#include <dmlc/logging.h>
+#include <set>
+#include <map>
+#include <mutex>
+#include <chrono>
+#include <iostream>
+#include <fstream>
+#include "./profiler.h"
+
+namespace mxnet {
+namespace engine {
+
+Profiler* Profiler::instance_ = nullptr;
+std::mutex Profiler::m_;
+
+Profiler::Profiler()
+    : state_(kNotRunning), enable_output_(false), mode_(kOnlySymbolic), filename_("profile.json") {
+  this->init_time_ = NowInUsec();
+
+  // TODO(ziheng) get device number during execution
+  int kMaxNumCpus = 60, kMaxNumGpus = 16;
+  this->cpu_num_ = kMaxNumCpus, this->gpu_num_ = kMaxNumGpus;
+
+  this->profile_stat = new DevStat[cpu_num_ + gpu_num_ + 1];
+  for (unsigned int i = 0; i < cpu_num_; ++i) {
+    profile_stat[i].dev_name = "cpu/" + std::to_string(i);
+  }
+  for (unsigned int i = 0; i < gpu_num_; ++i) {
+    profile_stat[cpu_num_ + i].dev_name = "gpu/" + std::to_string(i);
+  }
+  profile_stat[cpu_num_ + gpu_num_].dev_name = "cpu pinned/";
+}
+
+Profiler* Profiler::Get() {
+  std::lock_guard<std::mutex> lock{Profiler::m_};
```

Comment:
A lot of this can be written as lock free, ie use atomic_compare_exchange, to minimize distortion caused by profiler
we can merge the first version in and optimize later. but I think this is something we should keep in mind.